src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -973,6 +978,63 @@ func (c controller) waitingForClusterOperators(ctx context.Context) error {
 	return utils.WaitForPredicateWithTimer(ctxWithTimeout, WaitTimeout, GeneralProgressUpdateInt, isClusterVersionAvailable)
 }
 
+func areNodeLabelsUpdated(node v1.Node, nodeLabels string) bool {
+	nodeLabelsPresent := node.Labels
+	nodeLabelsRequired := make(map[string]string)
+	_ = json.Unmarshal([]byte(nodeLabels), &nodeLabelsRequired)
+
+	for key, value := range nodeLabelsRequired {
+		if val, found := nodeLabelsPresent[key]; !found || value != val {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (c *controller) updateNodesLabels() bool {
+	ignoreStatuses := []string{models.HostStatusDisabled, models.HostStatusError}
+	ctxReq := utils.GenerateRequestContext()
+	log := utils.RequestIDLogger(ctxReq, c.log)
+
+	assistedNodesMap, err := c.ic.GetHosts(ctxReq, log, ignoreStatuses)
+	if err != nil {
+		log.WithError(err).Error("Failed to get node map from the assisted service")
+		return KeepWaiting
+	}
+
+	nodes, err := c.kc.ListNodes()
+	if err != nil {
+		log.WithError(err).Error("Failed to get list of nodes from k8s client")
+		return KeepWaiting
+	} else if len(nodes.Items) != len(assistedNodesMap) {
```

Comment:
This is not good enough, (assisted) hosts that are in error may or may not appear as nodes in the nodes list, which will cause a discrepancy in the list lengths and fail this condition forever. Why have you decided to undo the "loop over the assistedNodesMap" that you had before?

Reply:
Can we not just remove the nodes from ListNodes which are in an error state before comparing the lengths?

Reply:
You could, but it's still feels like the wrong way to do it.

For example, another (admittedly obscure) case where you would still go wrong in your code even after the fix you just proposed, is the following (very stupid) DOS attack:

1. User starts installation
2. Installation is finalizing
3. Attacker uses the (public) worker.ign machine config to create nodes that try to join the cluster
4. This, if I recall correctly, creates a node resource with unapproved CSRs
5. Your code gets stuck because of length mismatch

The bottom line is that your goal is to look at the (non-error) hosts in the assisted API, find their corresponding nodes, and add the labels to those nodes. You're doing it the other way around (listing the nodes and finding their matching AI api hosts) it and makes these weird edge cases possible (needing to check for length match, having to remove error nodes from ListNodes, having to ignore external non-AI nodes that someone might add).